File: BOJ/2667.cpp
#include <cstdio>
#include <algorithm>
#include <vector>
#include <queue>
using namespace std;
int n,k;
int board[30][30];
bool visited[30][30];
int ans[900];
int gox[4] = {0,0,1,-1};
int goy[4] = {1,-1,0,0};
queue <pair<int, int>> q;
int bfs(int x, int y) {
	int cnt = 1;
	visited[x][y] = true;
	q.push({ x,y });
	while (!q.empty()) {
		int x = q.front().first;
		int y = q.front().second;
		q.pop();
		for (int i = 0; i < 4; i++) {
			int nx = x + gox[i];
			int ny = y + goy[i];
			if (nx<1 || nx>n || ny<1 || ny>n || visited[nx][ny] || board[nx][ny] == 0) continue;
			if (!visited[nx][ny] && board[nx][ny] == 1) {
				visited[nx][ny] = true;
				q.push({ nx,ny });
				cnt++;
			}
		}
	}
	return cnt;
}
int main(){
	scanf("%d", &n);
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			scanf("%1d", &board[i][j]);
		}
	}

	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			if (!visited[i][j] && board[i][j] == 1) {
				ans[k++]=bfs(i, j);
			}
		}
	}
	printf("%d\n", k);
	sort(ans, ans + k);
	for (int i = 0; i < k; i++) {
		printf("%d\n", ans[i]);
	}
	return 0;
}

File: BOJ(Python 공부용)/2667.py
n = int(input())
board = [[int(x) for x in input()] for y in range(n)]
visited = [[False] * n for _ in range(n)]
arr = []
cnt = 0


def dfs(x, y):
    global cnt
    visited[x][y] = True
    cnt += 1
    for dx, dy in (-1, 0), (1, 0), (0, -1), (0, 1):
        nx = x + dx
        ny = y + dy
        if nx < 0 or nx >= n or ny < 0 or ny >= n or board[nx][ny] == 0 or visited[nx][ny]:
            continue
        dfs(nx, ny)


for i in range(n):
    for j in range(n):
        if board[i][j] == 1 and not visited[i][j]:
            cnt = 0
            dfs(i, j)
            arr.append(cnt)

print(len(arr))
arr.sort()
for i in arr:
    print(i)
